File: pc/tests/timer.cpp
#include "timer.hpp"

using namespace std::chrono_literals;

int main() {

  static auto start = std::chrono::system_clock::now();
  static auto elapsed = []() -> float {
    return std::chrono::duration<float>(std::chrono::system_clock::now() - start).count();
  };

  espp::Logger logger({.tag = "Timer Test", .level = espp::Logger::Verbosity::INFO});

  logger.info("Starting timer test");

  espp::Timer timer({.name = "Timer",
                     .period = 0.100s,
                     .delay = 0.0s,
                     .callback = [&]() -> bool {
                       static int iterations = 0;
                       logger.info("[{:.3f}] timer fired {}", elapsed(), iterations++);
                       // don't want to stop the timer
                       return false;
                     },
                     .log_level = espp::Logger::Verbosity::DEBUG});

  std::this_thread::sleep_for(10s);

  logger.info("Stopping timer test");

  return 0;
}

File: python/timer.py
import time

from support_loader import espp

start = time.time()
def timer_func():
    global start
    elapsed = time.time() - start
    print(f"[{elapsed:.3f}] Hello from timer")
    return False # we don't want to stop the timer

timer = espp.Timer(espp.Timer.Config(
    "test timer", #name
    0.5, # period
    0.0, # delay
    timer_func,
    True, # auto start
    4096, # stack size (unused except on ESP)
    1, # priority (unused except on ESP)
    -1, # core (unused except on ESP)
    espp.Logger.Verbosity.none
))

time.sleep(5)

exit()
